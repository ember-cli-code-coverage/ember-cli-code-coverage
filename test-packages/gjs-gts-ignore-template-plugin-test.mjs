import gjsGtsIstanbulIgnoreTemplatePlugin from '../packages/ember-cli-code-coverage/lib/gjs-gts-istanbul-ignore-template-plugin';
import babel from '@babel/core';
import { expect, describe, it } from 'vitest';
import { Preprocessor } from 'content-tag';

let processor = new Preprocessor();

describe('gjs-gts-ignore-template-plugin', () => {
    it('it adds istanbul ignore comment before "template" invocation if template is imported from "@ember/template-compiler"', async () => {
        const example = processor.process(`import { on } from '@ember/modifier';

        let counter = 0;

        function bar() {
            counter++;
            console.log(counter);
        }

        <template>
          <button {{on "click" bar}}>
           increment
          </button>
        </template>`, { inline_source_map: true, filename: 'my-app/components/bar.gjs' });

        const {code} = babel.transform(example, {plugins: [gjsGtsIstanbulIgnoreTemplatePlugin], filename: 'some-file.gjs'});

        expect(code).toMatchSnapshot();
    });
  
    it('it does not add istanbul ignore comment before "template" invocation if template is not imported from "@ember/template-compiler"', async () => {
        const example = processor.process(`import { on } from '@ember/modifier';

        let counter = 0;

        function bar() {
            const template = ()=>{};
            template();
            counter++;
            console.log(counter);
        }

        <template>
          <button {{on "click" bar}}>
           increment
          </button>
        </template>`, { inline_source_map: true, filename: 'my-app/components/bar.gjs' });

        const {code} = babel.transform(example, {plugins: [gjsGtsIstanbulIgnoreTemplatePlugin], filename: 'some-file.gjs'});

        expect(code).toMatchSnapshot();
    });

    it('it skips files if it`s a regular js file', async () => {
        const example = processor.process(`import { on } from '@ember/modifier';

        let counter = 0;

        function bar() {
            counter++;
            console.log(counter);
        }

        <template>
          <button {{on "click" bar}}>
           increment
          </button>
        </template>`, { inline_source_map: true, filename: 'my-app/components/bar.js' });

        const {code} = babel.transform(example, {plugins: [gjsGtsIstanbulIgnoreTemplatePlugin], filename: 'some-file.js'});

        expect(code).toMatchSnapshot();
    });

    it('it skips files if it`s a regular ts file', async () => {
        const example = processor.process(`import { on } from '@ember/modifier';

        let counter = 0;

        function bar() {
            counter++;
            console.log(counter);
        }

        <template>
          <button {{on "click" bar}}>
           increment
          </button>
        </template>`, { inline_source_map: true, filename: 'my-app/components/bar.ts' });

        const {code} = babel.transform(example, {plugins: [gjsGtsIstanbulIgnoreTemplatePlugin], filename: 'some-file.ts'});

        expect(code).toMatchSnapshot();
    });
  });
  