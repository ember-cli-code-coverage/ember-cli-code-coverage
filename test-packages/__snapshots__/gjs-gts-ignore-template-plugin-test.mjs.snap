// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gjs-gts-ignore-template-plugin > it adds istanbul ignore comment before "template" invocation if template is imported from "@ember/template-compiler" 1`] = `
"import { template } from "@ember/template-compiler";
import { on } from '@ember/modifier';
let counter = 0;
function bar() {
  counter++;
  console.log(counter);
}
/* istanbul ignore next */
export default template(\`
          <button {{on "click" bar}}>
           increment
          </button>
        \`, {
  eval() {
    return eval(arguments[0]);
  }
});"
`;

exports[`gjs-gts-ignore-template-plugin > it does not add istanbul ignore comment before "template" invocation if template is not imported from "@ember/template-compiler" 1`] = `
"import { template as template1 } from "@ember/template-compiler";
import { on } from '@ember/modifier';
let counter = 0;
function bar() {
  const template1 = () => {};
  template1();
  counter++;
  console.log(counter);
}
/* istanbul ignore next */
export default template1(\`
          <button {{on "click" bar}}>
           increment
          </button>
        \`, {
  eval() {
    return eval(arguments[0]);
  }
});"
`;

exports[`gjs-gts-ignore-template-plugin > it skips files if it\`s a regular js file 1`] = `
"import { template } from "@ember/template-compiler";
import { on } from '@ember/modifier';
let counter = 0;
function bar() {
  counter++;
  console.log(counter);
}
export default template(\`
          <button {{on "click" bar}}>
           increment
          </button>
        \`, {
  eval() {
    return eval(arguments[0]);
  }
});"
`;

exports[`gjs-gts-ignore-template-plugin > it skips files if it\`s a regular ts file 1`] = `
"import { template } from "@ember/template-compiler";
import { on } from '@ember/modifier';
let counter = 0;
function bar() {
  counter++;
  console.log(counter);
}
export default template(\`
          <button {{on "click" bar}}>
           increment
          </button>
        \`, {
  eval() {
    return eval(arguments[0]);
  }
});"
`;
